[1/n] [c10d] Introduce a new TCPStore constructor

cbalioglu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58331 [4/n] [c10d] Introduce the multi-tenancy feature in TCPStore
* #58330 [3/n] [c10d] Revise the implementation of TCPStore
* #58329 [2/n] [c10d] Introduce the 'multiTenant' constructor parameter in TCPStore
* **#58328 [1/n] [c10d] Introduce a new TCPStore constructor**

This PR is part of a stack that addresses the GitHub issue #41614; it introduces a new `TCPStore` constructor that takes its optional parameters via a newly introduced `TCPStoreOptions` structure. This gives the API callers the flexibility to specify only the desired options while skipping the rest.

The main motivation behind this change is the introduction of the `multiTenant` constructor option in the second PR of this stack.

Differential Revision: [D28417742](https://our.internmc.facebook.com/intern/diff/D28417742/)

cbalioglu committed:
[1/n] [c10d] Introduce a new TCPStore constructor

This PR is part of a stack that addresses the GitHub issue #41614; it introduces a new `TCPStore` constructor that takes its optional parameters via a newly introduced `TCPStoreOptions` structure. This gives the API callers the flexibility to specify only the desired options while skipping the rest.

The main motivation behind this change is the introduction of the `multiTenant` constructor option in the second PR of this stack.

Differential Revision: [D28417742](https://our.internmc.facebook.com/intern/diff/D28417742/)

[ghstack-poisoned]

cbalioglu committed:
Update on "[1/n] [c10d] Introduce a new TCPStore constructor"


This PR is part of a stack that addresses the GitHub issue #41614; it introduces a new `TCPStore` constructor that takes its optional parameters via a newly introduced `TCPStoreOptions` structure. This gives the API callers the flexibility to specify only the desired options while skipping the rest.

The main motivation behind this change is the introduction of the `multiTenant` constructor option in the second PR of this stack.

Differential Revision: [D28417742](https://our.internmc.facebook.com/intern/diff/D28417742/)

[ghstack-poisoned]

rohan-varma commented on torch/lib/c10d/TCPStore.hpp:
```diff
@@ -124,12 +125,25 @@ class TCPStoreWorkerDaemon : public BackgroundThread {
   bool callbackRegisteredData_ = false;
 };
 
+struct TCPStoreOptions {
```
Curious: does it make sense to have a constructor versus the bracket-based initialization that's used? Not sure about the advantages/disadvantages there.

cbalioglu replied:
Since all fields have default values (meaning they have "optional" semantic) you do not gain anything by introducing a constructor. Aggregate initialization such as:

```
TCPStoreOptions opts{/* port */ 1234 };
opts.isServer = true;
```

is syntactically identical to what you would get with a constructor.

Beyond that the whole reason of an "options" struct is to avoid having a constructor with a long list of parameters.

rohan-varma reviewed (approved):
Looks great, thanks!

cbalioglu committed:
Update on "[1/n] [c10d] Introduce a new TCPStore constructor"


This PR is part of a stack that addresses the GitHub issue #41614; it introduces a new `TCPStore` constructor that takes its optional parameters via a newly introduced `TCPStoreOptions` structure. This gives the API callers the flexibility to specify only the desired options while skipping the rest.

The main motivation behind this change is the introduction of the `multiTenant` constructor option in the second PR of this stack.

Differential Revision: [D28417742](https://our.internmc.facebook.com/intern/diff/D28417742/)

[ghstack-poisoned]

cbalioglu committed:
Update on "[1/n] [c10d] Introduce a new TCPStore constructor"


This PR is part of a stack that addresses the GitHub issue #41614; it introduces a new `TCPStore` constructor that takes its optional parameters via a newly introduced `TCPStoreOptions` structure. This gives the API callers the flexibility to specify only the desired options while skipping the rest.

The main motivation behind this change is the introduction of the `multiTenant` constructor option in the second PR of this stack.

Differential Revision: [D28417742](https://our.internmc.facebook.com/intern/diff/D28417742/)

[ghstack-poisoned]

cbalioglu committed:
Update on "[1/n] [c10d] Introduce a new TCPStore constructor"


This PR is part of a stack that addresses the GitHub issue #41614; it introduces a new `TCPStore` constructor that takes its optional parameters via a newly introduced `TCPStoreOptions` structure. This gives the API callers the flexibility to specify only the desired options while skipping the rest.

The main motivation behind this change is the introduction of the `multiTenant` constructor option in the second PR of this stack.

Differential Revision: [D28417742](https://our.internmc.facebook.com/intern/diff/D28417742/)

[ghstack-poisoned]

cbalioglu committed:
Update on "[1/n] [c10d] Introduce a new TCPStore constructor"


This PR is part of a stack that addresses the GitHub issue #41614; it introduces a new `TCPStore` constructor that takes its optional parameters via a newly introduced `TCPStoreOptions` structure. This gives the API callers the flexibility to specify only the desired options while skipping the rest.

The main motivation behind this change is the introduction of the `multiTenant` constructor option in the second PR of this stack.

Differential Revision: [D28417742](https://our.internmc.facebook.com/intern/diff/D28417742/)

[ghstack-poisoned]

cbalioglu committed:
Update on "[1/n] [c10d] Introduce a new TCPStore constructor"


This PR is part of a stack that addresses the GitHub issue #41614; it introduces a new `TCPStore` constructor that takes its optional parameters via a newly introduced `TCPStoreOptions` structure. This gives the API callers the flexibility to specify only the desired options while skipping the rest.

The main motivation behind this change is the introduction of the `multiTenant` constructor option in the second PR of this stack.

Differential Revision: [D28417742](https://our.internmc.facebook.com/intern/diff/D28417742/)

[ghstack-poisoned]
